### PR TITLE
 Bump k8s dependencies to 1.20.4, csi spec to 1.3, external-snapshotter client to v4, go to 1.16

### PR DESCRIPTION
**Pull request by wongma7**
**Is this a bug fix or adding new feature?** fixes https://github.com/kubernetes-sigs/aws-ebs-csi-driver/issues/723

**What is this PR about? / Why do we need it?**

commits 1-5: go dependency bumps

commits 6-9: YAML image/API version bumps

commits 10-13: e2e test kops version bumps

**What testing is done?**

### CODE REVIEW

**Comment by wongma7** [comment]
client api changed, will fix.

 ../../../../pkg/mod/github.com/kubernetes-csi/external-snapshotter/v2@v2.0.1/pkg/client/clientset/versioned/typed/volumesnapshot/v1beta1/volumesnapshot.go:188:5: not enough arguments in call to c.client.Patch(pt).Namespace(c.ns).Resource("volumesnapshots").SubResource(subresources...).Name(name).Body(data).Do
	have ()
	want (context.Context)
../../../../pkg/mod/github.com/kubernetes-csi/external-snapshotter/v2@v2.0.1/pkg/client/clientset/versioned/typed/volumesnapshot/v1beta1/volumesnapshotclass.go:70:5: not enough arguments in call to c.client.Get().Resource("volumesnapshotclasses").Name(name).VersionedParams(&options, scheme.ParameterCodec).Do

**Comment by wongma7** [comment]
pkg/driver/driver.go:128:33: cannot use d (type *Driver) as type csi.ControllerServer in argument to csi.RegisterControllerServer:
	*Driver does not implement csi.ControllerServer (missing ControllerGetVolume method) 

I 'm going to leave it unimplemented

**Comment by wongma7** [comment]
The code added in 1.19 for getting metrics from kcm pod doesn't work for us:

Mar 15 16:56:01.625: INFO: ERROR unable to parse requirement: invalid label value: "kube-controller-manager-ip-172-20-39-70.us-west-2.compute.internal": at key: "name": must be no more than 63 characters

**Comment by wongma7** [comment]
Also, there is no way to disable this metrics grabbing code. We already check KCM metrics ourselves at the end of the test to see if csi or in tree was getting called so it's redundant :/

**Comment by wongma7** [comment]
hmmm I need to fix my logging code:


###
## Printing pod ebs-csi-controller-77fd584dc8-hkg6k ebs-plugin container logs
#
Error from server (Forbidden): Forbidden (user=system:anonymous, verb=get, resource=nodes, subresource=proxy) ( pods/log ebs-csi-controller-77fd584dc8-hkg6k)

**Comment by wongma7** [comment]
actually it's causing test failures as well since some of them check pod logs. 
```
                  s: "error running /usr/local/bin/kubectl --server=https://api-test-cluster-8868-k8s-rjdnu3-1175551248.us-west-2.elb.amazonaws.com --kubeconfig=/root/.kube/config --namespace=volume-3874 exec aws-injector --namespace=volume-3874 -- /bin/sh -c echo 'Hello from aws from namespace volume-3874' > /opt/0/index.html:\nCommand stdout:\n\nstderr:\nerror: unable to upgrade connection: Forbidden (user=system:anonymous, verb=create, resource=nodes, subresource=proxy)\n\nerror:\nexit status 1",
              }, 
```

**Comment by wongma7** [comment]
as of 119, kops kubelet default authorization mode is webook , probably it's cuz of that.

File: charts/aws-ebs-csi-driver/values.yaml [hunk 1]
```diff
@@ -18,7 +18,7 @@ sidecars:
     tag: "v3.0.0"
   snapshotterImage:
     repository: k8s.gcr.io/sig-storage/csi-snapshotter
-    tag: "v3.0.3"
+    tag: "v4.0.0"
```
Comment (ayberk) [inline charts/aws-ebs-csi-driver/values.yaml:21]: 4.0 requires k8s 1.20+. I don't think we should do this just yet.

https://kubernetes-csi.github.io/docs/external-snapshotter.html

File: charts/aws-ebs-csi-driver/values.yaml [hunk 2]
```diff
@@ -31,7 +31,7 @@ sidecars:
 
 snapshotController:
   repository: k8s.gcr.io/sig-storage/snapshot-controller
-  tag: "v3.0.3"
+  tag: "v4.0.0"
```
Comment (ayberk) [inline charts/aws-ebs-csi-driver/values.yaml:34]: Same for this one :( https://kubernetes-csi.github.io/docs/snapshot-controller.html

File: hack/e2e/kops.sh [hunk 3]
```diff
@@ -69,3 +78,36 @@ function kops_delete_cluster() {
   loudecho "Deleting cluster ${CLUSTER_NAME}"
   ${KOPS_BIN} delete cluster --name "${CLUSTER_NAME}" --state "${KOPS_STATE_FILE}" --yes
 }
+
+# TODO switch this to python, all this hacking with jq stinks!
```
Comment (ayberk) [inline hack/e2e/kops.sh:82]: +1

File: hack/e2e/kops.sh [hunk 4]
```diff
@@ -69,3 +78,36 @@ function kops_delete_cluster() {
...
+function kops_patch_cluster_file() {
+  CLUSTER_FILE=${1}
+  KOPS_PATCH_FILE=${2}
+
+  loudecho "Patching cluster $CLUSTER_NAME with $KOPS_PATCH_FILE"
+
+  # Temporary intermediate files for patching
+  CLUSTER_FILE_0=$CLUSTER_FILE.0
```
Comment (ayberk) [inline hack/e2e/kops.sh:90]: What does `$CLUSTER_FILE.0` do here?

**Review by ayberk — COMMENTED** [review]
👏 🙇

File: charts/aws-ebs-csi-driver/values.yaml (hunk 1 above)
Comment (wongma7) [reply charts/aws-ebs-csi-driver/values.yaml:21]: darn, ok

File: hack/e2e/kops.sh (hunk 4 above)
Comment (wongma7) [reply hack/e2e/kops.sh:90]: I can't open the file for writing via bash redirect and read it at the same time  to "patch" it hence this convoluted 'intermediate files' idea.

**Comment by wongma7** [comment]
uhh I think I accidentally pushed something to freak out the bot?? checking

**Comment by wongma7** [comment]
something to do with the blacklist /denylist change. https://github.com/kubernetes/test-infra/pull/21687

**Comment by wongma7** [comment]
TooManyLoadBalancers

File: deploy/kubernetes/base/controller.yaml [hunk 5]
```diff
@@ -32,7 +31,7 @@ spec:
           tolerationSeconds: 300
       containers:
         - name: ebs-plugin
-          image: k8s.gcr.io/provider-aws/aws-ebs-csi-driver:v0.10.0
+          image: k8s.gcr.io/provider-aws/aws-ebs-csi-driver:latest
```
Comment (ayberk) [inline deploy/kubernetes/base/controller.yaml:34]: Hmm this is kind of dangerous. What if we start building alpha images?

Comment (wongma7) [reply deploy/kubernetes/base/controller.yaml:34]: people should be using the stable overlay. However I agree best not to even mention latest tag, I'l lcheck where it's coming from (like is it  being propagated from a helm chart?)

**Comment by wongma7** [comment]
replaced by https://github.com/kubernetes-sigs/aws-ebs-csi-driver/pull/828

**Pull request closed without merging**